pkg/controller/persistentpodstate/persistent_pod_state_controller.go

```diff
@@ -0,0 +1,448 @@
+	for realReplicaCount := 0; realReplicaCount < int(sts.Replicas); replicaIndex++ {
+		if reserveOrdinals.Has(replicaIndex) {
+			continue
+		}
+		realReplicaCount++
+		replicas.Insert(replicaIndex)
+	}
+	return replicas.Has(index)
+}
+
+// auto generate PersistentPodState crd
+func (r *ReconcilePersistentPodState) syncPersistentPodState(req ctrl.Request) error {
+	// req.Name Format = sync#{apiVersion}#{workload.Kind}#{workload.Name}
+	// example for sync#apps/v1#StatefulSet#echoserver
+	arr := strings.Split(req.Name, "#")
+	if len(arr) != 4 {
+		klog.Warningf("Reconcile PersistentPodState workload(%s) is invalid", req.Name)
+		return nil
+	}
+	// fetch workload
+	apiVersion, kind, ns, name := arr[1], arr[2], req.Namespace, arr[3]
+	workload, err := r.finder.GetScaleAndSelectorForRef(apiVersion, kind, ns, name, "")
+	if err != nil {
+		return err
+	} else if workload == nil {
+		klog.Warningf("Reconcile PersistentPodState workload(%s) is Not Found", req.Name)
+		return nil
+	}
+	// fetch persistentPodState crd
+	persistentPodState := &appsv1alpha1.PersistentPodState{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: ns, Name: name}, persistentPodState)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// not found
+		persistentPodState = nil
+	}
+	val := workload.Metadata.Annotations[PersistentPodStateNodeTopology]
+	// create persistentPodState crd
+	if val == PersistentPodStatePreferredTopology || val == PersistentPodStateRequiredTopology {
+		// persistentPodState crd already exist
+		if persistentPodState != nil {
+			return nil
+		}
+
+		// create new persistentPodState crd
+		persistentPodState = &appsv1alpha1.PersistentPodState{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: ns,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: apiVersion,
+						Kind:       kind,
+						Name:       name,
+						Controller: utilpointer.BoolPtr(true),
+						UID:        workload.UID,
+					},
+				},
+			},
+			Spec: appsv1alpha1.PersistentPodStateSpec{
+				TargetReference: appsv1alpha1.TargetReference{
+					APIVersion: apiVersion,
+					Kind:       kind,
+					Name:       name,
+				},
+			},
+		}
+		if val == PersistentPodStatePreferredTopology {
+			persistentPodState.Spec.TopologyPreference = &appsv1alpha1.TopologyPreference{
+				Weight: 100,
+				Preference: appsv1alpha1.TopologyTerm{
+					TopologyKeys: []string{PodStateNodeTopologyLabel},
+				},
+			}
+		} else {
+			persistentPodState.Spec.TopologyConstraint = &appsv1alpha1.TopologyTerm{
+				TopologyKeys: []string{PodStateNodeTopologyLabel},
+			}
+		}
+
+		if err = r.Create(context.TODO(), persistentPodState); err != nil {
+			return err
+		}
+		klog.V(3).Infof("create StatefulSet(%s/%s) persistentPodState success", ns, name)
+		return nil
+	} else if persistentPodState == nil {
+		return nil
+	}
+
+	// delete persistentPodState crd
+	if err = r.Delete(context.TODO(), persistentPodState); err != nil {
```

Comment:
Should we delete the PersistentPodState object automatically? Maybe just let it be deleted by users or cascading delete if sts has been deleted.